sycl/test/basic_tests/buffer/buffer_ctad.cpp

```diff
@@ -0,0 +1,39 @@
+// RUN: %clangxx -std=c++17 -fsyntax-only -Xclang -verify %s
+// expected-no-diagnostics
+//==------------------- buffer_ctad.cpp - SYCL buffer CTAD test ----------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <CL/sycl.hpp>
+#include <cassert>
+#include <memory>
+
+using namespace cl::sycl;
+
+int main() {
+  std::vector<int> v(5, 1);
+  buffer b1(v.data(), range<1>(5));
+  static_assert(std::is_same_v<decltype(b1), buffer<int, 1>>);
+  buffer b1a(v.data(), range<1>(5), std::allocator<int>());
+  static_assert(
+      std::is_same_v<decltype(b1a), buffer<int, 1, std::allocator<int>>>);
+  buffer b1b(const_cast<const int *>(v.data()), range<1>(5));
+  static_assert(std::is_same_v<decltype(b1b), buffer<const int, 1>>);
```

Comment:
I think we don't want to deduce const here (same line 31 and 35). Rather it should just use the const overload constructor of the non-const type. This then means it only disables write-back but doesn't prevent writable accessor.